### ISSUE TITLE
CosmosDBInput returning 500 instead of 404 when item does not exist

### ISSUE THREAD

**Issue by ProTip** [Azure/azure-functions-dotnet-worker#2545]
### Description

I'm upgrading previously unmaintained application to the isolated hosting model. Unfortunately the `CosmosDBInput` binding is not behaving as expected. The following information has been pulled from the stack trace when attempting to access a missing resource.

The CosmosClient is receiving a 404 from the emulator:
```
Error:Microsoft.Azure.Cosmos.CosmosException : Response status code does not indicate success: NotFound (404)
```

However an exception is thrown trying to convert that response to the property type:
```
Error converting 1 input parameters for Function 'GetPreflightById': Cannot convert input parameter 'document' to type 'PreflightService.Models.PreflightRecord' from type 'Microsoft.Azure.Functions.Worker.Grpc.Messages.GrpcModelBindingData'
```

### Steps to reproduce

Try using `CosmosDBInput` to access a missing resource with the isolated worker model.

**Comment by kshyju** [comment]
This is because the cosmos input binding throws if the item is not found. Relevant code here:
https://github.com/Azure/azure-functions-dotnet-worker/blob/efd0aefe0a6564759f4aac57be8d6a7b8293ca86/extensions/Worker.Extensions.CosmosDB/src/CosmosDBConverter.cs#L114-L115

This is by design. If you want more control, you may consider binding to a `CosmosClient` instance which allow you to query and manage your data in any way you prefer. Please refer the samples [here](https://github.com/Azure/azure-functions-dotnet-worker/blob/efd0aefe0a6564759f4aac57be8d6a7b8293ca86/samples/Extensions/CosmosDB/CosmosInputBindingFunctions.cs).

**Comment by ProTip** [comment]
@kshyju I would like to challenge that this overall behavior is by design. I'm aware the binding throws, but are you suggesting the function returning a 500, internal server error, is _by design_ instead of an oversight?

If so this isn't mentioned in the [migration documentation](https://learn.microsoft.com/en-us/azure/azure-functions/migrate-cosmos-db-version-3-version-4?tabs=isolated-process&pivots=programming-language-csharp); it used to simply place `null` in the property.

The behavior is also not mentioned in the [binding documentation](https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-cosmosdb-v2-input?tabs=python-v2%2Cisolated-process%2Cnodejs-v4%2Cextensionv4&pivots=programming-language-csharp).

**Comment by bronthulke** [comment]
I agree, whether this was by design or not, whether the previous functionality was a mistake or not, it has changed and needs documenting on migration guides, as it's a breaking change.

Why not continue with the existing functionality so that the binding doesn't become useless for so many of us?

Also, we have had an open ticket about this for months with no feedback (see Protip's reference), can we please keep this conversation going since we are staring down the deadline for migration?

**Comment by bronthulke** [comment]
Do we need to take the lack of further discussion from Microsoft as a magical sign that this is indeed "by design", even though it was never documented and it has resulted in the binding becoming useless for a large number of use cases?

There are at least three tickets open about this (those are just the ones I'm personally aware of), and they are all still in limbo (as is most of our plans to migrate to the latest version of the extension).

**Comment by EntityAdam** [comment]
If we look at the example from the documentation here: https://learn.microsoft.com/en-us/azure/azure-functions/functions-bindings-cosmosdb-v2-input?tabs=python-v2%2Cisolated-process%2Cnodejs-v4%2Cextensionv4&pivots=programming-language-csharp#example

```cs
[Function(nameof(DocByIdFromJSON))]
public void DocByIdFromJSON(
    [QueueTrigger("todoqueueforlookup")] ToDoItemLookup toDoItemLookup,
    [CosmosDBInput(
        databaseName: "ToDoItems",
        containerName: "Items",
        Connection  = "CosmosDBConnection",
        Id = "{ToDoItemId}",
        PartitionKey = "{ToDoItemPartitionKeyValue}")] ToDoItem toDoItem)
{
    _logger.LogInformation($"C# Queue trigger function processed Id={toDoItemLookup?.ToDoItemId} Key={toDoItemLookup?.ToDoItemPartitionKeyValue}");

    if (toDoItem == null)
    {
        _logger.LogInformation($"ToDo item not found");
    }
    else
    {
        _logger.LogInformation($"Found ToDo item, Description={toDoItem.Description}");
    }
}
```

What I'm finding is that the `CosmosDbInput` will query Cosmos, which returns a `ResourceNotFound` exception, then the `CosmosDbInput` binding throws an exception making, which causes the function to never execute, so the following null check is actually unreachable code.

```cs
    if (toDoItem == null)
    {
        _logger.LogInformation($"ToDo item not found");
    }
```

Now there's a caveat, in the docs, it also states that this example is for the CosmosDb Bindings version 3.X.  For version 4.X, we only have a migration guide to go off of. I never tried the bindings version 3.X so I don't know if this is a behavior change or not.

I have created a reproduction of this here: https://github.com/EntityAdam/FunctionCrudSampleAspNet specifically the endpoint `/api/GetOne404Error`

**Comment by bronthulke** [comment]
Yes it is absolutely a behaviour change, that example flow works in 3.X. 

This is a big impact breaking change.

**Comment by florianlenz96** [comment]
If you're using version 4.3.0 of Microsoft.Azure.Functions.Worker.Extensions.CosmosDB, it works fine.
Switching from 4.3.0 to 4.4.0 adds this 500 problem

**Comment by mattchenderson** [comment]
We're assigning folks to review that PR. Thank you for opening it!

**Comment by kirankumarkolli** [comment]
Wrapping of the internal CosmosException into generic failure is happening at: 
https://github.com/Azure/azure-functions-dotnet-worker/blob/efd0aefe0a6564759f4aac57be8d6a7b8293ca86/extensions/Worker.Extensions.CosmosDB/src/CosmosDBConverter.cs#L63
